applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp

```diff
@@ -8,20 +8,166 @@
+    if ( dimension == 2 ) {
+        KRATOS_ERROR_IF( strain_size < 3 || strain_size > 4) << "Wrong constitutive law used. This is a 2D element! expected strain size is 3 or 4 (el id = ) " << pElement->Id() << std::endl;
+    } else {
+        KRATOS_ERROR_IF_NOT(strain_size == 6) << "Wrong constitutive law used. This is a 3D element! expected strain size is 6 (el id = ) "<<  pElement->Id() << std::endl;
+    }
+
+    return 0;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ComputeEquivalentF(
+    const Element* pElement,
+    Matrix& rF,
+    const Vector& rStrainTensor
+    )
+{
+    const auto& r_geometry = pElement->GetGeometry();
+    const SizeType dimension = r_geometry.WorkingSpaceDimension();
+
+    if(dimension == 2) {
+        rF(0,0) = 1.0+rStrainTensor(0);
+        rF(0,1) = 0.5*rStrainTensor(2);
+        rF(1,0) = 0.5*rStrainTensor(2);
+        rF(1,1) = 1.0+rStrainTensor(1);
+    } else {
+        rF(0,0) = 1.0+rStrainTensor(0);
+        rF(0,1) = 0.5*rStrainTensor(3);
+        rF(0,2) = 0.5*rStrainTensor(5);
+        rF(1,0) = 0.5*rStrainTensor(3);
+        rF(1,1) = 1.0+rStrainTensor(1);
+        rF(1,2) = 0.5*rStrainTensor(4);
+        rF(2,0) = 0.5*rStrainTensor(5);
+        rF(2,1) = 0.5*rStrainTensor(4);
+        rF(2,2) = 1.0+rStrainTensor(2);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void CalculateB(
+    const Element* pElement,
+    Matrix& rB,
+    const Matrix& rDN_DX
+    )
+{
+    const auto& r_geometry = pElement->GetGeometry();
+    const SizeType number_of_nodes = r_geometry.PointsNumber();
+    const SizeType dimension = r_geometry.WorkingSpaceDimension();
+
+    rB.clear();
+
+    if(dimension == 2) {
+        for ( SizeType i = 0; i < number_of_nodes; ++i ) {
+            rB(0, i*2    ) = rDN_DX(i, 0);
+            rB(1, i*2 + 1) = rDN_DX(i, 1);
+            rB(2, i*2    ) = rDN_DX(i, 1);
+            rB(2, i*2 + 1) = rDN_DX(i, 0);
+        }
+    } else if(dimension == 3) {
+        for ( SizeType i = 0; i < number_of_nodes; ++i ) {
+            rB(0, i*3    ) = rDN_DX(i, 0);
+            rB(1, i*3 + 1) = rDN_DX(i, 1);
+            rB(2, i*3 + 2) = rDN_DX(i, 2);
+            rB(3, i*3    ) = rDN_DX(i, 1);
+            rB(3, i*3 + 1) = rDN_DX(i, 0);
+            rB(4, i*3 + 1) = rDN_DX(i, 2);
+            rB(4, i*3 + 2) = rDN_DX(i, 1);
+            rB(5, i*3    ) = rDN_DX(i, 2);
+            rB(5, i*3 + 2) = rDN_DX(i, 0);
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+array_1d<double, 3> GetBodyForce(
```

Comment:
:+1:

I will hopefully use this too in the future :)

Do you think there is a solution for using this also for elements without integration-points (e.g. beam)?

Reply:
Yes, but should be updated first